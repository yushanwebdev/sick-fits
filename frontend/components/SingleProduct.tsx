/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { gql, useQuery } from '@apollo/client';
import Head from 'next/head';
import styled from 'styled-components';
import DisplayError from './DisplayError';

interface ISingleProductProps {
  id: string;
}

export const SINGLE_ITEM_QUERY = gql`
  query FindProduct($id: ID!) {
    Product(where: { id: $id }) {
      name
      description
      price
      id
      photo {
        image {
          publicUrlTransformed
        }
        altText
      }
    }
  }
`;

const ProductStyles = styled.div`
  display: grid;
  grid-auto-flow: column;
  grid-auto-columns: 1fr;
  max-width: var(--maxWidth);
  gap: 2rem;

  img {
    width: 100%;
    object-fit: contain;
  }
`;

export default function SingleProduct({ id }: ISingleProductProps) {
  const { loading, error, data } = useQuery(SINGLE_ITEM_QUERY, {
    variables: { id },
  });

  if (loading) {
    return <p>Loading...</p>;
  }

  if (error) {
    return <DisplayError error={error} />;
  }

  const { Product } = data;

  return (
    <ProductStyles>
      <Head>
        <title>Sick Fits | {Product.name}</title>
      </Head>
      <img
        src={Product.photo.image.publicUrlTransformed}
        alt={Product.photo.altText}
      />
      <div className="details">
        <h2>{Product.name}</h2>
        <p>{Product.description}</p>
      </div>
    </ProductStyles>
  );
}
