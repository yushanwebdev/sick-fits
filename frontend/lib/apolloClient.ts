// TODO: Fix the type issues
import { ApolloClient, from, InMemoryCache } from "@apollo/client";
import { onError } from "@apollo/client/link/error";
import { createUploadLink } from "apollo-upload-client";
import merge from "deepmerge";
import isEqual from "lodash/isEqual";
import { useMemo } from "react";
import { endpoint, prodEndpoint } from "../config";
import paginationField from "./paginationField";

export const APOLLO_STATE_PROP_NAME = "__APOLLO_STATE__";

let apolloClient;

const errorLink = onError(({ graphQLErrors, networkError }) => {
  if (graphQLErrors)
    graphQLErrors.forEach(({ message, locations, path }) =>
      console.log(
        `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
          locations
        )}, Path: ${JSON.stringify(path)}`
      )
    );
  if (networkError)
    console.log(`[Network error]: ${JSON.stringify(networkError)}`);
});

const uploadLink = createUploadLink({
  uri: process.env.NODE_ENV === "development" ? endpoint : prodEndpoint, // Server URL (must be absolute)
  fetchOptions: {
    credentials: "include", // To send the Cookie that stored in the Browser
  },
});

function createApolloClient() {
  return new ApolloClient({
    ssrMode: typeof window === "undefined",
    link: from([errorLink, uploadLink]),
    cache: new InMemoryCache({
      typePolicies: {
        Query: {
          fields: {
            // @ts-ignore
            allProducts: paginationField(),
          },
        },
      },
    }),
  });
}

export function initializeApollo(initialState = null) {
  const _apolloClient = apolloClient ?? createApolloClient();

  // If your page has Next.js data fetching methods that use Apollo Client, the initial state
  // gets hydrated here
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Merge the initialState from getStaticProps/getServerSideProps in the existing cache
    const data = merge(existingCache, initialState, {
      // combine arrays using object equality (like in sets)
      arrayMerge: (destinationArray, sourceArray) => [
        ...sourceArray,
        ...destinationArray.filter((d) =>
          sourceArray.every((s) => !isEqual(d, s))
        ),
      ],
    });

    // Restore the cache with the merged data
    _apolloClient.cache.restore(data);
  }
  // For SSG and SSR always create a new Apollo Client
  if (typeof window === "undefined") return _apolloClient;
  // Create the Apollo Client once in the client
  if (!apolloClient) apolloClient = _apolloClient;

  return _apolloClient;
}

export function addApolloState(client, pageProps) {
  try {
    if (pageProps?.props) {
      pageProps.props[APOLLO_STATE_PROP_NAME] = client.cache.extract() ?? null;
    }
  } catch (error) {
    console.log(error);
  }

  return pageProps;
}

export function useApollo(pageProps) {
  const state = pageProps[APOLLO_STATE_PROP_NAME];
  const store = useMemo(() => initializeApollo(state), [state]);
  return store;
}
