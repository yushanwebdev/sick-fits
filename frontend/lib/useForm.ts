import { ChangeEvent, useState } from 'react';

export default function useForm<T = Record<string, string>>(initial: T) {
  // create a state object for our inputs
  const [inputs, setInputs] = useState(initial);

  function handleChange(
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) {
    const { value, name, type } = e.target;
    let localValue: string | number | File = value;

    if (type === 'number') {
      localValue = parseInt(value);
    }

    if (type === 'file') {
      [localValue] = (e.target as HTMLInputElement).files;
    }

    setInputs({
      // copy the existing state
      ...inputs,
      [name]: localValue,
    });
  }

  function resetForm() {
    setInputs(initial);
  }

  function clearForm() {
    const blankState = Object.entries(inputs).reduce((acc, [key]) => {
      acc[key] = '';
      return acc;
    }, {} as T);

    setInputs(blankState);
  }

  // return the things we want to surface from this custom hook
  return {
    inputs,
    handleChange,
    resetForm,
    clearForm,
  };
}
