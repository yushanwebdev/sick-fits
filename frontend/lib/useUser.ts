/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-return */
import { gql, useQuery } from '@apollo/client';

export const CURRENT_USER_QUERY = gql`
  query GetAuthUser {
    authenticatedItem {
      ... on User {
        id
        email
        name
        cart {
          id
          quantity
          product {
            id
            price
            name
            description
            photo {
              image {
                publicUrlTransformed
              }
            }
          }
        }
      }
    }
  }
`;

export const useUser = () => {
  const { data, error } = useQuery(CURRENT_USER_QUERY);

  console.log('error', error);

  return data?.authenticatedItem;
};
