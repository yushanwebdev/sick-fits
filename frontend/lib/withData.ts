import {
  ApolloClient,
  InMemoryCache,
  from,
  NormalizedCacheObject,
  HttpOptions,
} from '@apollo/client';
import { onError } from '@apollo/client/link/error';
import { createUploadLink } from 'apollo-upload-client';
import withApollo from 'next-with-apollo';
import { getDataFromTree } from '@apollo/client/react/ssr';
import { endpoint, prodEndpoint } from '../config';

function createClient({
  headers,
  initialState,
}: {
  headers: HttpOptions['headers'];
  initialState: NormalizedCacheObject;
}) {
  const errorLink = onError(({ graphQLErrors, networkError }) => {
    if (graphQLErrors)
      graphQLErrors.forEach(({ message, locations, path }) =>
        console.log(
          `[GraphQL error]: Message: ${message}, Location: ${JSON.stringify(
            locations
          )}, Path: ${JSON.stringify(path)}`
        )
      );

    if (networkError)
      console.log(`[Network error]: ${JSON.stringify(networkError)}`);
  });

  // this uses apollo-link-http under the hood, so all the options here come from that package
  const uploadLink = createUploadLink({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    credentials: 'include',
    // pass the headers along from this request. This enables SSR with logged in state
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    headers,
  });

  // If you provide a link chain to ApolloClient, you
  // don't provide the `uri` option.
  return new ApolloClient({
    // The `from` function combines an array of individual links
    // into a link chain
    link: from([errorLink, uploadLink]),
    cache: new InMemoryCache().restore(initialState || {}),
  });
}

export default withApollo(createClient, { getDataFromTree });
