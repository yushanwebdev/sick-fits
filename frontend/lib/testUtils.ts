import casual from "casual";
import { PAGINATION_QUERY } from "../components/Pagination";

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
  __typename: "Item",
  id: "abc123",
  price: 5000,
  user: null,
  photo: {
    __typename: "Photo",
    id: "abc123",
    altText: "dogs are best",
    image: {
      __typename: "Image",
      publicUrlTransformed: "dog.jpg",
    },
  },
  name: "dogs are best",
  description: "dogs",
});

const fakeUser = (overrides?: Object) => ({
  __typename: "User",
  id: "4234",
  name: casual.name,
  email: casual.email,
  permissions: ["ADMIN"],
  orders: [],
  cart: [],
  role: {
    canManageCart: true,
    canManageUsers: true,
    canManageProducts: true,
  },
  ...overrides,
});

const fakeOrderItem = () => ({
  __typename: "OrderItem",
  id: casual.uuid,
  image: {
    image: `${casual.word}.jpg`,
  },
  name: casual.words(),
  price: 4234,
  quantity: 1,
  description: casual.words(),
});

const fakeOrder = () => ({
  __typename: "Order",
  id: "ord123",
  charge: "ch_123",
  total: 40000,
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: "2022-12-11T20:16:13.797Z",
  user: fakeUser(),
});

const fakeCartItem = (overrides?: Object) => ({
  __typename: "CartItem",
  id: "omg123",
  quantity: 3,
  product: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

function makePaginationMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allProductsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: "aggregate",
            aggregate: {
              count: length,
              __typename: "count",
            },
          },
        },
      },
    },
  ];
}

export {
  makePaginationMocksFor,
  fakeItem,
  fakeUser,
  fakeCartItem,
  fakeOrder,
  fakeOrderItem,
};
