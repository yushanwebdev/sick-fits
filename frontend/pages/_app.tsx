/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/ban-ts-comment */
import { ApolloProvider } from '@apollo/client';
import type { AppProps } from 'next/app';
import { Router } from 'next/router';
import styled, { ThemeProvider, DefaultTheme } from 'styled-components';
import NProgress from 'nprogress';
import Header from '../components/Header';
import GlobalStyle from '../components/styles/GlobalStyles';
import { useApollo } from '../lib/apolloClient';

import '../components/styles/nprogress.css';

const theme: DefaultTheme = {
  colors: {
    red: '#ff0000',
    black: '#393939',
    grey: '#3a3a3a',
    get gray() {
      return (this as DefaultTheme['colors']).grey;
    },
    lightGrey: '#e1e1e1',
    get lightGray() {
      return (this as DefaultTheme['colors']).lightGrey;
    },
    offWhite: '#ededed',
    maxWidth: '1000px',
    bs: '0 12px 24px 0 rgba(0, 0, 0, 0.09)',
  },
};

const InnerStyles = styled.div`
  max-width: var(--maxWidth);
  margin: 0 auto;
  padding: 2rem;
`;

Router.events.on('routeChangeStart', () => NProgress.start());
Router.events.on('routeChangeComplete', () => NProgress.done());
Router.events.on('routeChangeError', () => NProgress.done());

const MyApp = ({ Component, pageProps }: AppProps) => {
  const apolloClient = useApollo(pageProps);

  return (
    <ApolloProvider client={apolloClient}>
      <ThemeProvider theme={theme}>
        <GlobalStyle />
        <Header />
        <InnerStyles>
          <Component {...pageProps} />
        </InnerStyles>
      </ThemeProvider>
    </ApolloProvider>
  );
};

export default MyApp;
