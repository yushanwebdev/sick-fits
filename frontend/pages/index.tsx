import { ALL_PRODUCTS_QUERY } from "../components/Products";
import { perPage } from "../config";
import { addApolloState, initializeApollo } from "../lib/apolloClient";
import { CURRENT_USER_QUERY } from "../lib/useUser";
import ProductsPage from "./products";

export async function getServerSideProps() {
  try {
    const apolloClient = initializeApollo();

    await apolloClient.query({
      query: ALL_PRODUCTS_QUERY,
      variables: {
        skip: 0,
        first: perPage,
      },
    });

    await apolloClient.query({
      query: CURRENT_USER_QUERY,
    });

    return addApolloState(apolloClient, {
      props: {},
    });
  } catch (error) {
    console.log(error);
  }
}

export default ProductsPage;
