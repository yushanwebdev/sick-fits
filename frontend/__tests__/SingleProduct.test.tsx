import { MockedProvider } from "@apollo/client/testing"
import { render, screen } from "@testing-library/react"
import { ThemeProvider } from "styled-components"
import SingleProduct, { SINGLE_ITEM_QUERY } from "../components/SingleProduct"
import { fakeItem } from "../lib/testUtils"
import { theme } from "../pages/_app"

const product = fakeItem()

const mocks = [
    {
        // When someone requests this query and variable combo
        request: {
            query: SINGLE_ITEM_QUERY,
            variables: {
                id: '123'
            }
        },
        // Return this data
        result: {
            data: {
                Product: product
            }
        }
    }
]

describe('<SingleProduct />', () => {
    it('renders with proper data', async () => {
        // We need to make some fake data
        const {container, debug} = render(
            <MockedProvider mocks={mocks}><ThemeProvider theme={theme}><SingleProduct id="123" /></ThemeProvider></MockedProvider>
        )
        
        // Wait for the test ID to show up
        await screen.findByTestId('singleProduct')
        expect(container).toMatchSnapshot()
    })

    it('Errors out when an item is not found', async () => {
        const errorMock = [
            {
                request: {
                    query: SINGLE_ITEM_QUERY,
                    variables: {
                        id: '123'
                    }
                },
                error: {
                    name: 'Error',
                    message: 'Item not found!!!'
                }
            }
        ]

        const {container, debug} = render(
            <MockedProvider mocks={errorMock}><ThemeProvider theme={theme}><SingleProduct id="123" /></ThemeProvider></MockedProvider>
        )

        
        // Wait for the test ID to show up
        await screen.findByTestId('graphqlError') // always pass an test id in camelCase if not you will get errors
        expect(container).toHaveTextContent('Shoot!')
        expect(container).toHaveTextContent('Item not found!!!')
    })
})