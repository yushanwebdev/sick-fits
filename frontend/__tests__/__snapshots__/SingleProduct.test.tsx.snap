// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleProduct /> renders with proper data 1`] = `
<div>
  <div
    class="sc-d96fd322-0 eaNrYj"
    data-testid="singleProduct"
  >
    <img
      alt="dogs are best"
      src="dog.jpg"
    />
    <div
      class="details"
    >
      <h2>
        dogs are best
      </h2>
      <p>
        dogs
      </p>
    </div>
  </div>
</div>
`;
